### PR TITLE
RIOT-OS support with GNRC TCP/IP sockets

### PR DESCRIPTION
**Pull request by danielinux**
Add support for RIOT-OS GNRC TCP/IP sockets via WOLFSSL_GNRC

### CODE REVIEW

File: wolfssl/wolfio.h [hunk 1]
```diff
@@ -491,6 +492,37 @@ WOLFSSL_API void wolfSSL_SetIOWriteFlags(WOLFSSL* ssl, int flags);
...
 
+#ifdef WOLFSSL_GNRC
+    #include <sock_types.h>
+    #include <net/gnrc.h>
+    #include <net/af.h>
+    #include <net/sock.h>
+    #include <net/gnrc/tcp.h>
+    #include <net/gnrc/udp.h>
+
+    struct gnrc_wolfssl_ctx {
```
Comment (toddouska) [inline wolfssl/wolfio.h]: Does this structure need to be in a header?  That is, if it's only used internally by wolfio.c couldn't it go in the .c file?

File: wolfssl/wolfio.h [hunk 2]
```diff
@@ -491,6 +492,37 @@ WOLFSSL_API void wolfSSL_SetIOWriteFlags(WOLFSSL* ssl, int flags);
...
+        WOLFSSL_CTX *ctx;
+        WOLFSSL *ssl;
+
+        int closing;
+        struct _sock_tl_ep peer_addr;
+    };
+
+    typedef struct gnrc_wolfssl_ctx sock_tls_t;
+
+    WOLFSSL_LOCAL int GNRC_Receive(WOLFSSL* ssl, char* buf, int sz,
```
Comment (toddouska) [inline wolfssl/wolfio.h]: If the visibility of these callbacks is WOLFSSL_LOCAL how would a user get access to them to install them if they're not setup automatically?

File: src/wolfio.c [hunk 3]
```diff
@@ -2225,4 +2225,78 @@ int uIPGenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
...
+ * return : bytes sent, or error
+ */
+
+int GNRC_SendTo(WOLFSSL* ssl, char* buf, int sz, void* _ctx)
+{
+    sock_tls_t *ctx = (sock_tls_t *)_ctx;
+    int ret = 0;
+    (void)ssl;
+    ret = sock_udp_send(&ctx->conn.udp, (unsigned char *)buf, sz, &ctx->peer_addr);
+    if (ret <= 0)
```
Comment (toddouska) [inline src/wolfio.c]: sent 0 (can't do any bytes now) is different than <0 (some type of error).  wolfSSL will interpret this possible error state as 0 and continue to retry sending.

File: src/wolfio.c [hunk 4]
```diff
@@ -2225,4 +2225,78 @@ int uIPGenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
...
+ */
+int GNRC_GenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
+{
+    sock_tls_t *ctx = (sock_tls_t *)_ctx;
+    byte token[32];
+    byte digest[WC_SHA_DIGEST_SIZE];
+    int  ret = 0;
+    (void)ssl;
+    XMEMSET(token, 0, sizeof(token));
+    XMEMCPY(token, &ctx->peer_addr, sizeof(sock_udp_ep_t));
```
Comment (toddouska) [inline src/wolfio.c]: What if the sizeof sock_udp_ep_t becomes bigger than 32?

File: src/wolfio.c [hunk 5]
```diff
@@ -2225,4 +2225,78 @@ int uIPGenerateCookie(WOLFSSL* ssl, byte *buf, int sz, void *_ctx)
...
+ *  return : nb bytes read, or error
+ */
+int GNRC_Receive(WOLFSSL *ssl, char *buf, int sz, void *_ctx)
+{
+    sock_udp_ep_t ep;
+    int ret;
+    uint32_t timeout = wolfSSL_dtls_get_current_timeout(ssl) * 1000000;
+    sock_tls_t *ctx = (sock_tls_t *)_ctx;
+    if (!ctx)
+        return -1;
```
Comment (toddouska) [inline src/wolfio.c]: WOLFSSL_CBIO_ERR_GENERAL would be better here.

File: wolfssl/wolfio.h (hunk 1 above)
Comment (danielinux) [reply wolfssl/wolfio.h]: This is the structure shared with RIOT-OS and exported (for the wolfssl RIOT-OS package) as `sock_tls_t`. See the API from our proposed wolfssl pkg in RIOT-OS:
https://github.com/danielinux/RIOT/blob/master/pkg/wolfssl/sock_tls/sock_tls.h

File: wolfssl/wolfio.h (hunk 2 above)
Comment (danielinux) [reply wolfssl/wolfio.h]: now they are set automatically as suggested by @kaleb-himes

File: src/wolfio.c (hunk 3 above)
Comment (danielinux) [reply src/wolfio.c]: Fixed, thanks

File: src/wolfio.c (hunk 4 above)
Comment (danielinux) [reply src/wolfio.c]: Fixed with a more strict check on the allocated size. Removed hardcoded number.

**Comment by danielinux** [comment]
@kaleb-himes : implemented all the changes you proposed:

1. Only send/recv are supported /needed (Renamed GNRC_SendTo to GNRC_Send).
2. went for suggested option 'B'
3. I've done that, and associated the callback automatically. However, associating the `gnrc_wolfssl_ctx` to the internal pointer `gnrcCtx` after `wolfSSL_new()` does not modify the callback context (the tls_sock will still have to call `wolfSSL_SetIOReadCtx` and `wolfSSL_SetIOWriteCtx` after `wolfSSL_new`.

**Comment by kaleb-himes** [comment]
Daniel,

Changes are great, I really apologize for being nit-picky on this one item, the send/recv. For traditional TCP/IP stacks the ```send()``` ```recv()``` are used, for UDP (datagram protocol) ```sentTo()``` ```recvFrom()``` are used.

Since this is DTLS can we use the same underlying UDP conventions by calling them ```GNRC_RecvFrom``` and ```GNRC_SendTo```? No other changes are necessary, you are assigning them to the appropriate ctx element in ```ctx->CBIORecv = GNRC_Receive``` I am only being nit-picky about the naming IE ```ctx->CBIORecv = GNRC_RecvFrom``` would be preferred. When it's named RecvFrom that automatically tells those of our developers who sometimes work on the DTLS code that they are DTLS (UDP) callbacks and not TLS (TCP) callbacks.

KH

**Comment by kaleb-himes** [comment]
Over to @toddouska for final review.

**Pull request merged**